pkg/storaged/cockpit-clevis-tool.py

```diff
@@ -0,0 +1,162 @@
+    'RSA': ['kty', 'p', 'd', 'q', 'dp', 'dq', 'qi', 'oth'],
+    'EC':  ['kty', 'crv', 'x', 'y'],
+    'oct': ['kty', 'k'],
+}
+
+def compute_thp(jwk):
+    jwk = {k: v for k, v in jwk.items() if k in REQUIRED_ATTRS[jwk['kty']]}
+    jwk = json.dumps(jwk, sort_keys=True, separators=(',', ':'))
+    thp = hashlib.sha1(jwk.encode('utf8')).digest()
+    return base64.urlsafe_b64encode(thp).decode('utf8').rstrip('=')
+
+def is_signing_key(jwk):
+    use = jwk.get('use', None)
+    ops = jwk.get('key_ops', None)
+    if use is None and ops is None:
+        return True
+    if use == 'sig':
+        return True
+    if ops is not None and 'verify' in ops:
+        return True
+    return False
+
+def get_adv(url):
+    if not "://" in url:
+        url = "http://" + url
+    with urllib.request.urlopen(url + "/adv") as rsp:
+        adv_text = rsp.read()
+        adv = json.loads(adv_text.decode())
+        payload = json.loads(b64_decode(adv['payload']))
+        info = {
+            "adv": adv,
+            "sha1sum": hashlib.sha1(adv_text).hexdigest(),
+            "keys": list(map(compute_thp, payload['keys'])),
+            "sigkeys": list(map(compute_thp, filter(is_signing_key, payload['keys']))),
+        }
+        sys.stdout.write(json.dumps(info) + "\n")
+        sys.stdout.flush()
+
+def add(dev, url, adv, passphrase):
+    config = { "url": url, "adv": json.loads(adv) }
+    subprocess.run([ "clevis", "luks", "bind", "-f", "-k", "-", "-d", dev, "tang", json.dumps(config) ],
+                   check=True, input=passphrase.encode())
+
+def check_key(dev, slot):
+    jwe = subprocess.check_output([ "luksmeta", "load", "-d", dev, "-s", slot ])
+    subprocess.run([ "clevis", "decrypt" ],
+                   check = True, input = jwe, stdout=subprocess.PIPE).stdout
+
+def replace(dev, slot, url, adv):
+    jwe = subprocess.check_output([ "luksmeta", "load", "-d", dev, "-s", slot ])
+    passphrase = subprocess.run([ "clevis", "decrypt" ],
+                                check = True, input = jwe, stdout=subprocess.PIPE).stdout
+    if len(adv) > 0:
+        config = { "url": url, "adv": json.loads(adv) }
+    else:
+        config = { "url": url }
+    new_jwe = subprocess.run([ "clevis", "encrypt", "tang", json.dumps(config) ],
+                             check = True, input = passphrase, stdout=subprocess.PIPE).stdout
+    subprocess.run([ "luksmeta", "wipe", "-d", dev, "-s", slot, "-f" ])
+    subprocess.run([ "luksmeta", "save", "-d", dev, "-s", slot, "-u", "cb6e8904-81ff-40da-a84a-07ab9ab5715e" ],
+                   check = True, input = new_jwe)
+
+def unlock(dev):
+    # XXX - clevis-luks-unlock always exits 1, so we check whether the
+    # expected device exists afterwards.
+    clear_dev = b"luks-" + subprocess.check_output([ "cryptsetup", "luksUUID", dev ]).strip()
+    subprocess.run([ "clevis", "luks", "unlock", "-d", dev, "-n", clear_dev ])
+    sys.exit(0 if os.path.exists(b"/dev/mapper/" + clear_dev) else 1)
```

Comment:
This is guaranteed to be synchronous, i. e. `clevis luks unlock` always waits until the cleartext device appears? Or might this need a waiting loop?

Reply:
It's synchronous, thank god.